clang/lib/CIR/Dialect/IR/CIRDialect.cpp

```diff
@@ -2132,6 +2133,117 @@ LogicalResult cir::ComplexImagPtrOp::verify() {
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// Bit manipulation operations
+//===----------------------------------------------------------------------===//
+
+template <typename F>
+static OpFoldResult foldUnaryBitOp(mlir::Attribute inputAttr, F func,
+                                   bool poisonZero = false) {
+  auto input = mlir::dyn_cast_if_present<IntAttr>(inputAttr);
+  if (!input)
+    return nullptr;
+
+  llvm::APInt inputValue = input.getValue();
+  if (poisonZero && inputValue.isZero()) {
+    // TODO(cir): maybe we should return a poison value here?
```

Comment:
This seems like a reasonable thing to do. If we generate the corresponding intrinsic during lowering, the optimizer will fold it to poison, so we may as well do that here.

Reply:
Added a missing feature guard here since `PoisonAttr` is not in the upstream yet. Will add in later patches.